CanvasRenderingContext2DBase::getImageData with ImageDataStorageFormat

squelart opened the pull request:
#### d4570c2ebeca6f19cf3016be1679e5f9aa3a74ee
<pre>
CanvasRenderingContext2DBase::getImageData with ImageDataStorageFormat
<a href="https://bugs.webkit.org/show_bug.cgi?id=284414">https://bugs.webkit.org/show_bug.cgi?id=284414</a>
<a href="https://rdar.apple.com/problem/141249363">rdar://problem/141249363</a>

Reviewed by Cameron McCormack.

`context.getImageData(..., {storageFormat:&quot;float16&quot;})` now works and
returns a Float16Array, which is currently converted from the canvas&apos;
always-uint8 buffer.

`TypedArrayItemConverter&lt;Float16Array, Uint8ClampedArray&gt;::convert`
tweaked to round towards nearest (so 0.5 gets converted to 128), as
per specs <a href="https://github.com/w3c/ColorWeb-CG/blob/main/canvas_float.md">https://github.com/w3c/ColorWeb-CG/blob/main/canvas_float.md</a>

* LayoutTests/fast/canvas/imagedata-storageformat-enabled-expected.txt:
* LayoutTests/fast/canvas/imagedata-storageformat-enabled.html:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
(WebCore::ImageData::ImageData):
* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::create):
* Source/WebCore/html/ImageDataArray.cpp:
(WebCore::ImageDataArray::ImageDataArray):
(WebCore::ImageDataArray::extractBufferViewWithStorageFormat):
* Source/WebCore/html/ImageDataArray.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::getImageData const):

Canonical link: <a href="https://commits.webkit.org/289717@main">https://commits.webkit.org/289717@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34886941784d08837e1f8563ef572108b9f8940d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87759 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7275 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42145 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92624 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38509 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89810 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15446 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67757 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25502 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90761 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5845 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79404 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48126 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5632 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33810 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37616 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76030 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34689 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94510 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14927 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10968 "Found 1 new test failure: imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-svg-script-element.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76606 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15182 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75260 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75842 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18651 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20212 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18643 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7915 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14943 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20246 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14687 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18131 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/html/ImageDataArray.cpp:
```diff
@@ -184,6 +188,18 @@ Ref<JSON::Value> ImageDataArray::copyToJSONArray() const
+    switch (*overridingStorageFormat) {
+    case ImageDataStorageFormat::Uint8: return asUint8ClampedArray();
+    case ImageDataStorageFormat::Float16: return asFloat16Array();
+    default: RELEASE_ASSERT_NOT_REACHED("Unexpected ImageDataStorageFormat value");
```
Nit: slightly better to put the assert just below the `switch`, so that we can rely on the compiler telling us if there is a new, valid enum value that we haven't handled yet.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2512,17 +2512,22 @@ RefPtr<ByteArrayPixelBuffer> CanvasRenderingContext2DBase::cacheImageDataIfPossi
 }
 
-RefPtr<ImageData> CanvasRenderingContext2DBase::makeImageDataIfContentsCached(const IntRect& sourceRect, PredefinedColorSpace colorSpace) const
+RefPtr<ImageData> CanvasRenderingContext2DBase::makeImageDataIfContentsCached(const IntRect& sourceRect, PredefinedColorSpace colorSpace, ImageDataStorageFormat storageFormat) const
```
I think it would be fine to make this caching mechanism work only for Uint8 storage, and then not worry about passing the storage format value in.

squelart committed:
CanvasRenderingContext2DBase::getImageData with ImageDataStorageFormat
https://bugs.webkit.org/show_bug.cgi?id=284414
rdar://problem/141249363

Reviewed by Cameron McCormack.

`context.getImageData(..., {storageFormat:"float16"})` now works and
returns a Float16Array, which is currently converted from the canvas'
always-uint8 buffer.

`TypedArrayItemConverter<Float16Array, Uint8ClampedArray>::convert`
tweaked to round towards nearest (so 0.5 gets converted to 128), as
per specs https://github.com/w3c/ColorWeb-CG/blob/main/canvas_float.md

* LayoutTests/fast/canvas/imagedata-storageformat-enabled-expected.txt:
* LayoutTests/fast/canvas/imagedata-storageformat-enabled.html:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
(WebCore::ImageData::ImageData):
* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::create):
* Source/WebCore/html/ImageDataArray.cpp:
(WebCore::ImageDataArray::ImageDataArray):
(WebCore::ImageDataArray::extractBufferViewWithStorageFormat):
* Source/WebCore/html/ImageDataArray.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::getImageData const):

Canonical link: https://commits.webkit.org/289717@main
